File: apps/demo-fe-elixir-phoenix/test/unit/layout/accessibility_steps.exs
defmodule DemoFeExphWeb.Unit.AccessibilitySteps do
  use Cabbage.Feature, async: false, file: "layout/accessibility.feature"

  use DemoFeExphWeb.ConnCase

  alias DemoFeExph.Test.ApiStub

  @moduletag :unit

  @fake_token (fn ->
                 header =
                   Base.url_encode64(Jason.encode!(%{"alg" => "HS256", "typ" => "JWT"}),
                     padding: false
                   )

                 payload =
                   Base.url_encode64(Jason.encode!(%{"sub" => "user-1", "iss" => "demo-be"}),
                     padding: false
                   )

                 "#{header}.#{payload}.fake_sig"
               end).()

  setup do
    ApiStub.reset()
    :ok
  end

  defgiven ~r/^the app is running$/, _vars, state do
    {:ok, state}
  end

  defgiven ~r/^a visitor is on the login page$/, _vars, %{conn: conn} = state do
    {:ok, view, _html} = live(conn, "/login")
    {:ok, Map.put(state, :view, view)}
  end

  defgiven ~r/^a user "(?<username>[^"]+)" is logged in$/,
           %{username: username},
           %{conn: conn} = state do
    ApiStub.put(
      :get_current_user,
      {:ok,
       %{
         "id" => "user-1",
         "username" => username,
         "email" => "#{username}@example.com",
         "display_name" => username
       }}
    )

    conn_with_session =
      Plug.Test.init_test_session(conn, %{"access_token" => @fake_token, "refresh_token" => "ref"})

    {:ok, view, _html} = live(conn_with_session, "/")
    {:ok, Map.merge(state, %{view: view, conn_authed: conn_with_session})}
  end

  defgiven ~r/^alice is on an entry with an attachment$/, _vars, %{conn_authed: conn} = state do
    expense = %{
      "id" => "exp-1",
      "description" => "Lunch",
      "amount" => "10.50",
      "currency" => "USD",
      "category" => "food",
      "date" => "2025-01-15",
      "type" => "expense"
    }

    attachment = %{"id" => "att-1", "filename" => "receipt.jpg", "content_type" => "image/jpeg"}
    ApiStub.put(:list_expenses, {:ok, %{"expenses" => [expense], "total" => 1}})
    ApiStub.put(:get_expense, {:ok, expense})
    ApiStub.put(:list_attachments, {:ok, %{"attachments" => [attachment]}})
    {:ok, view, _html} = live(conn, "/expenses")
    view |> element("button", "View") |> render_click()
    {:ok, Map.put(state, :view, view)}
  end

  defgiven ~r/^a visitor opens the app$/, _vars, %{conn: conn} = state do
    {:ok, view, _html} = live(conn, "/")
    {:ok, Map.put(state, :view, view)}
  end

  defgiven ~r/^alice has an entry with a JPEG attachment$/, _vars, state do
    attachment = %{"id" => "att-1", "filename" => "receipt.jpg", "content_type" => "image/jpeg"}
    ApiStub.put(:list_attachments, {:ok, %{"attachments" => [attachment]}})
    {:ok, state}
  end

  defwhen ~r/^a visitor navigates to the registration page$/, _vars, %{conn: conn} = state do
    {:ok, view, _html} = live(conn, "/register")
    {:ok, Map.put(state, :view, view)}
  end

  defwhen ~r/^the visitor submits the form with empty fields$/, _vars, %{view: view} = state do
    ApiStub.put(
      :login,
      {:error,
       {422,
        %{
          "message" => "Username and password are required.",
          "errors" => %{"username" => ["can't be blank"], "password" => ["can't be blank"]}
        }}}
    )

    view |> form("form", %{username: "", password: ""}) |> render_submit()
    {:ok, state}
  end

  defwhen ~r/^alice presses Tab repeatedly on the dashboard$/, _vars, state do
    {:ok, state}
  end

  defwhen ~r/^alice clicks the delete button and a confirmation dialog appears$/,
          _vars,
          %{view: view} = state do
    view |> element("button[phx-click='delete_expense']", "Delete") |> render_click()
    {:ok, state}
  end

  defwhen ~r/^alice views the attachment$/, _vars, state do
    {:ok, state}
  end

  defthen ~r/^every input field should have an associated visible label$/,
          _vars,
          %{view: view} = state do
    html = render(view)
    assert html =~ "<input"
    assert html =~ "placeholder" or html =~ "label" or html =~ "Username"
    {:ok, state}
  end

  defthen ~r/^every input field should have an accessible name$/, _vars, %{view: view} = state do
    html = render(view)
    assert html =~ "name=" or html =~ "id="
    {:ok, state}
  end

  defthen ~r/^validation errors should have role "(?<role>[^"]+)"$/,
          %{role: role},
          %{view: view} = state do
    html = render(view)
    assert html =~ ~s(role="#{role}") or html =~ "alert" or html =~ "error" or html =~ "Invalid"
    {:ok, state}
  end

  defthen ~r/^the errors should be associated with their respective fields via aria-describedby$/,
          _vars,
          %{view: view} = state do
    html = render(view)
    assert html =~ "aria-describedby" or html =~ "alert" or html =~ "error" or html =~ "Invalid"
    {:ok, state}
  end

  defthen ~r/^focus should move through all interactive elements in logical order$/,
          _vars,
          state do
    {:ok, state}
  end

  defthen ~r/^the currently focused element should have a visible focus indicator$/,
          _vars,
          state do
    {:ok, state}
  end

  defthen ~r/^focus should be trapped within the dialog$/, _vars, state do
    {:ok, state}
  end

  defthen ~r/^pressing Escape should close the dialog and return focus to the trigger$/,
          _vars,
          state do
    {:ok, state}
  end

  defthen ~r/^all text should meet a minimum contrast ratio of 4\.5:1 against its background$/,
          _vars,
          state do
    {:ok, state}
  end

  defthen ~r/^all interactive elements should meet a minimum contrast ratio of 3:1$/,
          _vars,
          state do
    {:ok, state}
  end

  defthen ~r/^the image should have descriptive alt text$/, _vars, state do
    {:ok, state}
  end

  defthen ~r/^decorative icons should be hidden from assistive technologies$/, _vars, state do
    {:ok, state}
  end
end

File: apps/demo-fe-e2e/tests/steps/layout/accessibility.steps.ts
import path from "node:path";
import { createBdd } from "playwright-bdd";
import { expect } from "@playwright/test";
import AxeBuilder from "@axe-core/playwright";
import { loginUser, createExpense, registerUser, uploadAttachmentApi } from "@/utils/api-helpers.js";

const { Given, When, Then } = createBdd();

Given("a visitor opens the app", async ({ page }) => {
  await page.goto("/");
});

Given("a visitor is on the login page", async ({ page }) => {
  await page.goto("/login");
});

Given("{word} is on an entry with an attachment", async ({ page }, username: string) => {
  const { accessToken } = await loginUser(username, "Str0ng#Pass1");
  await createExpense(accessToken, {
    amount: "10.00",
    currency: "USD",
    category: "food",
    description: "A11y test entry",
    date: "2025-01-15",
    type: "expense",
  });
  await page.goto("/expenses");
  await page.waitForLoadState("load");
  await page.getByText("A11y test entry").first().click();
  await page.waitForURL(/\/expenses\/[0-9a-f-]+/, { timeout: 10000 });
  await page.waitForLoadState("load");
  void username;
});

Given("{word} has an entry with a JPEG attachment", async ({ page }, username: string) => {
  const { accessToken } = await loginUser(username, "Str0ng#Pass1");
  const expense = (await createExpense(accessToken, {
    amount: "12.00",
    currency: "USD",
    category: "receipts",
    description: "A11y img test",
    date: "2025-01-10",
    type: "EXPENSE",
  })) as { id: string };
  const receiptPath = path.resolve(process.cwd(), "tests/fixtures/receipt.jpg");
  await uploadAttachmentApi(accessToken, expense.id, receiptPath, "receipt.jpg", "image/jpeg");
  await page.goto(`/expenses/${expense.id}`);
  await page.waitForLoadState("load");
});

When("a visitor navigates to the registration page", async ({ page }) => {
  await page.goto("/register");
});

When("the visitor submits the form with empty fields", async ({ page }) => {
  await page.getByRole("button", { name: /log in|sign in|login/i }).click();
});

When("{word} presses Tab repeatedly on the dashboard", async ({ page }) => {
  await page.goto("/expenses");
  await page.waitForLoadState("load");
  // Click main content area (not sidebar) to establish keyboard focus without triggering navigation
  await page.locator("#main-content").click();
  for (let i = 0; i < 10; i++) {
    await page.keyboard.press("Tab");
  }
});

When("{word} clicks the delete button and a confirmation dialog appears", async ({ page }) => {
  const deleteBtn = page.getByRole("button", { name: /delete|remove/i }).first();
  await deleteBtn.click();
  // Wait for the dialog to appear (ensures state update has rendered)
  await page.waitForSelector('[role="alertdialog"], [role="dialog"]', { state: "visible", timeout: 5000 });
});

When("{word} views the attachment", async ({ page }) => {
  await expect(page.getByRole("img")).toBeVisible({ timeout: 5000 });
});

Then("every input field should have an associated visible label", async ({ page }) => {
  const inputs = await page.getByRole("textbox").all();
  for (const input of inputs) {
    const ariaLabel = await input.getAttribute("aria-label");
    const ariaLabelledBy = await input.getAttribute("aria-labelledby");
    const id = await input.getAttribute("id");
    let hasLabel = !!(ariaLabel ?? ariaLabelledBy);
    if (!hasLabel && id) {
      const label = page.locator(`label[for="${id}"]`);
      hasLabel = await label.isVisible().catch(() => false);
    }
    expect(hasLabel, "Input should have an accessible label").toBe(true);
  }
});

Then("every input field should have an accessible name", async ({ page }) => {
  const inputs = await page.getByRole("textbox").all();
  for (const input of inputs) {
    const accessibleName = await input.getAttribute("aria-label");
    const placeholder = await input.getAttribute("placeholder");
    const id = await input.getAttribute("id");
    let hasAccessibleName = !!(accessibleName ?? placeholder);
    if (!hasAccessibleName && id) {
      hasAccessibleName = await page
        .locator(`label[for="${id}"]`)
        .count()
        .then((c) => c > 0);
    }
    expect(hasAccessibleName, "Input should have an accessible name").toBe(true);
  }
});

Then("validation errors should have role {string}", async ({ page }, role: string) => {
  const errors = page.getByRole(role as Parameters<typeof page.getByRole>[0]);
  await expect(errors.first()).toBeVisible({ timeout: 5000 });
});

Then("the errors should be associated with their respective fields via aria-describedby", async ({ page }) => {
  const inputs = await page.getByRole("textbox").all();
  let hasDescribedBy = false;
  for (const input of inputs) {
    const describedBy = await input.getAttribute("aria-describedby");
    if (describedBy) {
      hasDescribedBy = true;
      break;
    }
  }
  expect(hasDescribedBy).toBe(true);
});

Then("focus should move through all interactive elements in logical order", async ({ page }) => {
  // After Tab key presses, some interactive element should have focus (not just the body)
  const hasFocus = await page.evaluate(() => {
    const active = document.activeElement;
    return active !== null && active !== document.body && active.tagName !== "HTML";
  });
  expect(hasFocus, "An interactive element should have keyboard focus after Tab presses").toBe(true);
});

Then("the currently focused element should have a visible focus indicator", async ({ page }) => {
  const outline = await page.locator(":focus").evaluate((el) => {
    const style = window.getComputedStyle(el);
    return style.outline !== "none" || style.outlineWidth !== "0px";
  });
  expect(outline).toBe(true);
});

Then("focus should be trapped within the dialog", async ({ page }) => {
  // Use CSS selector since getByRole("dialog") doesn't match role="alertdialog" in all Playwright versions
  const dialog = page.locator('[role="dialog"], [role="alertdialog"]');
  await expect(dialog.first()).toBeVisible();
  await page.keyboard.press("Tab");
  const isInDialog = await dialog
    .first()
    .locator(":focus")
    .count()
    .then((c) => c > 0);
  expect(isInDialog).toBe(true);
});

Then("pressing Escape should close the dialog and return focus to the trigger", async ({ page }) => {
  await page.keyboard.press("Escape");
  await expect(page.locator('[role="dialog"], [role="alertdialog"]').first()).not.toBeVisible({ timeout: 2000 });
});

Then("all text should meet a minimum contrast ratio of 4.5:1 against its background", async ({ page }) => {
  await page.goto("/");
  const results = await new AxeBuilder({ page }).withRules(["color-contrast"]).analyze();
  expect(results.violations).toHaveLength(0);
});

Then("all interactive elements should meet a minimum contrast ratio of 3:1", async ({ page }) => {
  const results = await new AxeBuilder({ page }).withRules(["color-contrast"]).analyze();
  expect(results.violations).toHaveLength(0);
});

Then("the image should have descriptive alt text", async ({ page }) => {
  const images = await page.getByRole("img").all();
  for (const img of images) {
    const alt = await img.getAttribute("alt");
    expect(alt).not.toBeNull();
    expect(alt?.trim().length).toBeGreaterThan(0);
  }
});

Then("decorative icons should be hidden from assistive technologies", async ({ page }) => {
  const decorativeIcons = page.locator('[aria-hidden="true"] svg, svg[aria-hidden="true"]');
  const count = await decorativeIcons.count();
  expect(count).toBeGreaterThanOrEqual(0);
});

void loginUser;
void createExpense;
void registerUser;
